kong/concurrency.lua

```diff
@@ -51,7 +51,8 @@ function concurrency.with_worker_mutex(opts, fn)
   local elapsed, err = rlock:lock(opts_name)
   if not elapsed then
     if err == "timeout" then
-      return nil, err
+      local ttl = rlock.dict and rlock.dict:ttl(opts_name)
```

Comment:
I'm having hard time understanding what `ttl` means in context of timed out mutex? TTL stands for Time To Live right? If a mutex timeout out I'd assumed that TTL reached 0 or something? I see that a bit lower you rephrase it as `retry_after`. Is that the idea behind TTL - to notify the user when a timed out mutex call should be retried?

Reply:
see: https://github.com/openresty/lua-resty-lock?tab=readme-ov-file#new

- timeout is how long to wait before giving up while trying to obtain the lock
- exptime is the lifetime of the lock entry, after which the lock is always released

ttl refers to the latter, this is basically (an attempt at) rewriting the logic [here](https://github.com/Kong/kong/blob/3eeb9c94fda11a2a8afe34d7b2ef8d032d0a6ff3/kong/db/declarative/import.lua#L581-L582) but using `lua-resty-lock` instead of the shared dictionary directly. Notice that lua-resty-lock [never returns the "exists" error](https://github.com/openresty/lua-resty-lock/blob/405d0bf4cbfa74d742c6ed3158d442221e6212a9/lib/resty/lock.lua#L127-L177), in case of "exists" and if a timeout occurs, it returns "timeout" (in which case here we read the ttl, i.e. the time left until exptime) and return it [to be used as you said to "retry after"](https://github.com/Kong/kong/blob/3eeb9c94fda11a2a8afe34d7b2ef8d032d0a6ff3/kong/api/routes/config.lua#L151-L157).

Reply:
Thank you! Now I get it 😄 !